linera-chain/src/certificate.rs

```diff
@@ -0,0 +1,304 @@
+        }
+    }
+}
+
+impl<T> CertificateT<T> {
+    pub fn new(
+        value: T,
+        old_hash: CryptoHash,
+        round: Round,
+        mut signatures: Vec<(ValidatorName, Signature)>,
+    ) -> Self {
+        signatures.sort_by_key(|&(validator_name, _)| validator_name);
+        Self {
+            value: Hashed::unchecked_new(value, old_hash),
+            round,
+            signatures,
+        }
+    }
+
+    pub fn signatures(&self) -> &Vec<(ValidatorName, Signature)> {
+        &self.signatures
+    }
+
+    /// Adds a signature to the certificate's list of signatures
+    /// It's the responsibility of the caller to not insert duplicates
+    pub fn add_signature(
+        &mut self,
+        signature: (ValidatorName, Signature),
+    ) -> &Vec<(ValidatorName, Signature)> {
+        let index = self
+            .signatures
+            .binary_search_by(|(name, _)| name.cmp(&signature.0))
+            .unwrap_or_else(std::convert::identity);
+        self.signatures.insert(index, signature);
+        &self.signatures
+    }
+
+    /// Returns whether the validator is among the signatories of this certificate.
+    pub fn is_signed_by(&self, validator_name: &ValidatorName) -> bool {
+        self.signatures
+            .binary_search_by(|(name, _)| name.cmp(validator_name))
+            .is_ok()
+    }
+}
+
+impl<T> CertificateT<T> {
+    /// Returns the certified value's hash.
+    pub fn hash(&self) -> CryptoHash {
+        self.value.hash()
+    }
+
+    /// Verifies the certificate.
+    pub fn check(&self, committee: &Committee) -> Result<(), ChainError> {
+        crate::data_types::check_signatures(self.hash(), self.round, &self.signatures, committee)?;
+        Ok(())
+    }
+}
+
+impl ValidatedBlockCertificate {
+    pub fn requires_blob(&self, blob_id: &BlobId) -> bool {
+        self.executed_block().requires_blob(blob_id)
+    }
+
+    #[cfg(with_testing)]
+    pub fn outgoing_message_count(&self) -> usize {
+        self.executed_block().messages().iter().map(Vec::len).sum()
+    }
+
+    /// Returns the `LiteValue` corresponding to the certified value.
+    pub fn lite_value(&self) -> LiteValue {
+        LiteValue {
+            value_hash: self.value.hash(),
+            chain_id: self.executed_block().block.chain_id,
+        }
+    }
+
+    /// Returns the certificate without the full value.
+    pub fn lite_certificate(&self) -> LiteCertificate<'_> {
+        LiteCertificate {
+            value: self.lite_value(),
+            round: self.round,
+            signatures: Cow::Borrowed(&self.signatures),
+        }
+    }
+
+    /// Returns reference to the `ExecutedBlock` contained in this certificate.
+    pub fn executed_block(&self) -> &ExecutedBlock {
+        self.inner().inner()
+    }
+}
+
+impl From<Certificate> for ValidatedBlockCertificate {
+    fn from(cert: Certificate) -> Self {
+        let signatures = cert.signatures().clone();
+        let hash = cert.value.hash();
```

Comment:
Should there just be a `Hashed::new` method that would do the hashing? I get that `unchecked_new` could be needed in places where we already have the hash and do not want to waste CPU time on recomputing it, but in this case it looks like having `Hashed` compute the hash might be a good idea.

Reply:
I need a hash of the "old" type (here `Certificate`) so I cannot recompute it over the `ValidatedBlockCertificate` one.

I need the hash of the old type b/c signature is done over the old hash. Validators/clients would fail.